### PR TITLE
Feat/2432 pagination sqltable

### LINKED ISSUE #2432: Add pagination to sql_table
### Feature description

The goal would be to add the possibility to paginate queries, with a defined page size with a logic like this:

1. Get query and make sure there's a key to sort by
2. Modify the query to set the page size by adding a LIMIT
3. Execute the query and retrieve results
4. IF rows returned < page size, add OFFSET of page size and repeat previous step

This could also be updated for parallel executions, simply by executing a count and then dividing the load among the other workers. As an improvement, the pages would overlap to avoid skipping any rows by mistake, then the result can flow through the pipeline.

### Are you a dlt user?

Yes, I'm already a dlt user.

### Use case

When using sql_table and sql_database with databases that have limitations on the amount of rows retrieved per transaction, dlt does not currently offer the possibility of paginating through the query with pagination (LIMIT & OFFSET). There is the possibility to make changes to the DB, but due to technical restrictions from the production environment this might not be feasible.

### Proposed solution

Add the possibility to paginate SQL query imports to bypass these restrictions without requiring changes on the source database.

### Related issues

_No response_

### PR DESCRIPTION
**Pull request by lucargir**
### Description
Add support for pagination to the SQL source

### Related Issues
- Closes #2432

I have followed the instructions regarding linting, formatting and tests and so far everything was fine. Let me know if there's anything that I have missed or should be changed.

### CODE REVIEW

**Comment by lucargir** [comment]
@rudolfix Fixed the formatting issue due to the last commit, make lint now returns 0 issues.

File: dlt/sources/sql_database/pagination_helpers.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        logger.warning(f"Retrying query with backoff. New page size {self._page_size}")
+        self._current_retry += 1
+        # Update query
+        self._query = self._query.limit(self._page_size)
+        return True
+
+    def make_query(self) -> CursorResult:
+        self._query = self._query.offset(self._query_offset())
+        try:
+            result = self._conn.execute(self._query)
```
Comment (rudolfix) [inline dlt/sources/sql_database/pagination_helpers.py:73]: you probably do not want this code in production but to discover page size and then use it without retries, right?

**Review by rudolfix — CHANGES_REQUESTED** [review]
code looks pretty good! but I think we can simplify the pagination. we also lack proper test: this feature is relatively complicated and we should test different combinations of last value func (min, max), row order (ASC, DESC) to make sure we are not leaking any data, we have such parametrized tests and we are glad to help once we have code finalized

File: dlt/sources/sql_database/pagination_helpers.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        self._n_rows: int = self._total_rows(conn=conn, query=query)
+        self._total_pages: int = ceil(self._n_rows / self._page_size)
+        self._query = query.limit(self._page_size)
+        if pk_columns:
+            self._query = self._query.order_by(*pk_columns)
+        logger.info(
+            "Initial pagination query:"
+            f" {self._query.compile(compile_kwargs={'literal_binds': True})}"
+        )
+        self._max_retries = coerce_value(
```
[the remark this answers is not in the document]
Comment (lucargir) [reply dlt/sources/sql_database/pagination_helpers.py:34]: Actually, that was my reasoning at first, but I couldn't figure out a way to get the number of rows from the CursorResult, just found that it has a row_count attribute, so I'll update it.

File: dlt/sources/sql_database/pagination_helpers.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        query: SelectClause,
+        conn: Connection,
+        page_size: int,
+        pk_columns: Optional[list[ColumnElement[Any]]] = None,
+    ) -> None:
+        if not isinstance(query, Select):
+            raise TypeError("The query adapter is currently not compatible with pagination")
+        self._page_size: int = page_size
+        self._conn = conn
+        self._current_page: int = 0
```
[the remark this answers is not in the document]
Comment (lucargir) [reply dlt/sources/sql_database/pagination_helpers.py:24]: Actually, that was my reasoning at first, but I couldn't figure out a way to get the number of rows from the CursorResult, just found that it has a row_count attribute, so I'll update it.

File: dlt/sources/_core_source_templates/sql_database_pipeline.py [hunk 4]
```diff
@@ -9,9 +9,13 @@
 
 from dlt.sources.sql_database import sql_database, sql_table, Table
 
+from dlt.pipeline.exceptions import PipelineStepFailed
+
 from sqlalchemy.sql.sqltypes import TypeEngine
 import sqlalchemy as sa
 
+from pytest import raises
```
[the remark this answers is not in the document]
Comment (lucargir) [reply dlt/sources/_core_source_templates/sql_database_pipeline.py:17]: Will move this to an external testing for the pagination helper, I mixed things up.

File: dlt/sources/sql_database/helpers.py [hunk 5]
```diff
@@ -104,6 +106,11 @@ def __init__(
             self.on_cursor_value_missing = None
             self.range_start = None
             self.range_end = None
+            if page_size:
```
[the remark this answers is not in the document]
Comment (lucargir) [reply dlt/sources/sql_database/helpers.py:109]: The only issue I see here is that if we don't sort by the primary key and the cursor, we might get non-deterministic results, given that the cursor might be repeated among rows. For example:

id, user, sign_up_date
1, john, 20250101
2, mary, 20250101
3, jane, 20250103

With a page size of 1, sorting by sign_up_date, you could get different results each time the query is executed. And the current implementation allows for only one cursor column, that's why I added the sorting by PK, because it should be unique, and therefore PK + cursor should be deterministic.

File: dlt/sources/sql_database/pagination_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucargir) [reply dlt/sources/sql_database/pagination_helpers.py:34]: This was meant more for debugging when developing the pipeline. The query might fail for some unexpected reason, and I though that giving the opportunity to get the raw query sent to the DB could be useful, hence the info level of the log.

[the remark this answers is not in the document]
Comment (lucargir) [reply dlt/sources/sql_database/pagination_helpers.py]: If you were referring to the max_retries, this comment explains the reason behind it:
https://github.com/dlt-hub/dlt/pull/2477#discussion_r2036720118

**Comment by lucargir** [comment]
Okay I made some changes:
✅ Updated the docstrings for sql_table and sql_database
✅ Added docstrings for the paginator
✅ Removed the pytest from the pipeline template
✅ Removed the initial query count, rely on number of rows returned after each query is executed
❓Removed arbitrary sorting order for PK, rely on row_order
❓Remove additional sorting by PK when PK and cursor are the same

Regarding the parametrised tests, I assume you mean those from tests/load/sources/sql_database? I would need some help on how to approach it. Also, I think adding some test to make sure that all rows are returned without any duplicates would be a good idea, but also not sure what would be the best approach.

File: dlt/sources/sql_database/helpers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rudolfix) [reply dlt/sources/sql_database/helpers.py:125]: you are right! our `Incremental` class does this kind of deduplication by requesting overlapping values but here indeed we must have fully deterministic order

**Review by rudolfix — COMMENTED** [review]
@lucargir we had internal discussion on this and we see a lot of value in what you did. but your case is still pretty specific and contains details that may not work or perform poorly on other backends that yours
- your extend ORDER BY to include primary key. in many cases this will prevent the index on incremental to be used, optimizer will pick index on PK which may result in full table scan
- you use OFFSET to do the pagination. so all records below offset will be read and then discarded
- you use very specific backoff mechanism with a handler on Exception that trigger it. so it is very unspecific
- you still use your own configuration via env variables

and I think it is fine and it solves your problem.

my take is that we would let `TableLoader` to be subclassed and refactor and standardize internal interface ie. `_load_rows` so `TablePaginator` that you wrote could be easily plugged in.

we would document this. we could also link to your `TablePaginator` from our docs as an example. WDYT?

**Comment by lucargir** [comment]
@rudolfix Yup, it completely makes sense. Regarding the OFFSET and the ORDER BY, I don't think there's any other option to guarantee the determinism. For the latter, it's more because I don't know how to approach it, so:
- Backoff: this could be set-up as a config rather than an env variable, where the user decides whether it should be used or not
- Exception: I could change it to OperationalError, so it catches only those, but I am not sure I can be specific to memory issues.

Nonetheless, I think that what you suggested with TableLoader would be the best. This would allow for more flexibility, and specific cases like mine could be easily handled without requiring a feature request. If there's anything I could help with, I'd be happy to contribute with documenting, or making the PR for the custom TableLoader when creating a sql_database/sql_table.

**Comment by rudolfix** [comment]
as mentioned we are closing this PR in favor of implementing https://github.com/dlt-hub/dlt/issues/2736. thanks for contributing this. tbh. I can't give you an ETA when TableLoader will be abstracted in the current code...
